login/config.c

```diff
@@ -101,48 +103,182 @@ static void key_value(char *line, char **key, char **val) {
+  l = strtol(val, &endptr, 0);
+  if (errno) {
+    return status_createf("ERROR: invalid numeric value: %s", val);
+  }
+  if (*endptr != '\0' || l < 0 || l > UINT_MAX) {
+    return status_createf("ERROR: invalid value for option: %s", val);
+  }
+  *option = (unsigned int)l;
+  return STATUS_OK;
+}
+
+static status_t set_bitfield_option(glome_login_config_t *config, uint8_t bit) {
+  config->options |= bit;
+  return STATUS_OK;
+}
+
+static status_t clear_bitfield_option(glome_login_config_t *config,
+                                      uint8_t bit) {
+  config->options &= ~bit;
+  return STATUS_OK;
+}
+
+static bool boolean_true(const char *val) {
+  if (strcasecmp(val, "true") == 0) {
+    return true;
+  } else if (strcasecmp(val, "yes") == 0) {
+    return true;
+  } else if (strcasecmp(val, "on") == 0) {
+    return true;
+  } else if (strcmp(val, "1") == 0) {
     return true;
   }
+  return false;
+}
 
-  if (strcmp(key, "key") == 0) {
-    if (is_zeroed(config->service_key, sizeof config->service_key)) {
-      if (decode_hex(config->service_key, sizeof config->service_key, val)) {
-        errorf("ERROR: Failed to hex decode service key\n");
-        return false;
-      }
-    }
-  } else if (strcmp(key, "key-version") == 0) {
-    char *end;
-    errno = 0;
-    long n = strtol(val, &end, 10);
-    if (errno != 0 || *end != '\0') {
-      errorf("ERROR: Failed to parse service key version\n");
-      return false;
-    }
-    if (n <= 0) {
-      errorf("ERROR: Key version should be a positive value\n");
-      return false;
+static bool boolean_false(const char *val) {
+  if (strcasecmp(val, "false") == 0) {
+    return true;
+  } else if (strcasecmp(val, "no") == 0) {
+    return true;
+  } else if (strcasecmp(val, "off") == 0) {
+    return true;
+  } else if (strcmp(val, "0") == 0) {
+    return true;
+  }
+  return false;
+}
+
+static status_t update_bitfield_option(glome_login_config_t *config,
+                                       uint8_t bit, bool invert,
+                                       const char *val) {
+  if (boolean_true(val)) {
+    if (invert) {
+      return clear_bitfield_option(config, bit);
+    } else {
+      return set_bitfield_option(config, bit);
     }
-    if (n > 255) {
-      errorf("ERROR: Key version too large, must fit into 8-bit int\n");
-      return false;
+  } else if (boolean_false(val)) {
+    if (invert) {
+      return set_bitfield_option(config, bit);
+    } else {
+      return clear_bitfield_option(config, bit);
     }
-    if (n > 0 && config->service_key_id == 0) {
-      config->service_key_id = n;
+  } else {
+    return status_createf("ERROR: unrecognized boolean value: %s", val);
+  }
+}
+
+static status_t assign_key_option(uint8_t *dest, size_t dest_len,
+                                  const char *val) {
+  if (is_zeroed(dest, dest_len)) {
+    if (decode_hex(dest, dest_len, val)) {
+      return status_createf("ERROR: Failed to hex decode service key: %s", val);
```

Comment:
We don't seem to be consistent in `ERROR: <uppercase>` vs. `ERROR: <lowercase>`.

Reply:
Done.